pydantic/main.py

```diff
@@ -246,7 +244,24 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__repr__': Representation.__repr__,
             **{n: v for n, v in namespace.items() if n not in fields},
         }
-        return super().__new__(mcs, name, bases, new_namespace, **kwargs)
+        cls = super().__new__(mcs, name, bases, new_namespace, **kwargs)
+        if _custom_root_type:
+            _setup_custom_root_type(cls)
+        return cls
+
+
+def _setup_custom_root_type(cls: Type['Model']) -> None:
+    original_parse_obj = cls.parse_obj
+
+    root_field = cls.__fields__[ROOT_KEY]
+    root_field_is_mapping = root_field.shape == SHAPE_MAPPING
+
+    def parse_obj(obj: Any) -> 'Model':
+        if not isinstance(obj, dict) or root_field_is_mapping:
```

Comment:
By checking specifically for `root_field_is_mapping`, we ensure there are no changes to code paths that were previously valid.

-----

This check could be done inside the current `parse_obj` function, and that would allow us to drop the `_setup_custom_root_type` function. But the use of this `_setup_custom_root_type` function ensures that models *without* a custom root type are never forced to experience any root-type-checking overhead.

@samuelcolvin I'm expecting you to be opposed to this on the grounds of the complexity it introduces, but figured there was a chance you might like the pattern for the performance benefits. I am happy to unwind it and just put the checks inside the `parse_obj` function where it is defined if that's preferable.